src/core/bpf-firewall.c

```diff
@@ -453,6 +796,85 @@ static int bpf_firewall_prepare_access_maps(
         return 0;
 }
 
+static int bpf_firewall_prepare_port_maps(
+                Unit *u,
+                int verdict,
+                int *ret_port_map_ingress_fd,
+                int *ret_port_map_egress_fd) {
+
+        size_t n_ingress_ports = 0, n_egress_ports = 0;
+        Unit *p;
+        int r;
+
+        assert(ret_port_map_ingress_fd);
+        assert(ret_port_map_egress_fd);
+
+        for (p = u; p; p = UNIT_DEREF(p->slice)) {
+                CGroupContext *cc;
+
+                cc = unit_get_cgroup_context(p);
+                if (!cc)
+                        continue;
+
+                bpf_firewall_count_port_range_access_items(verdict == ACCESS_ALLOWED ? cc->port_ingress_allow : cc->port_ingress_deny, &n_ingress_ports);
+                bpf_firewall_count_port_range_access_items(verdict == ACCESS_ALLOWED ? cc->port_egress_allow : cc->port_egress_deny, &n_egress_ports);
+        }
+
+        if (n_ingress_ports > 0) {
+                r = bpf_map_new(BPF_MAP_TYPE_LPM_TRIE,
```

Comment:
hmm, is LPM_TRIE really the best map for this? maybe there are other more suitable ones for storing IP port range data?

Reply:
I thought about this a bit, but didn't find any really good alternatives. One option would be to add all ports in a range as single items to a hash map, but the map would get big very soon. Another would be to do some data structure logic in the BPF filter, but I would like to keep that as the "fast path" and leave the data structure code to kernel.